src/main/java/com/exactpro/th2/crawler/state/RecoveryState.java

```diff
@@ -0,0 +1,300 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exactpro.th2.crawler.state;
+
+import com.exactpro.cradle.Direction;
+import com.exactpro.cradle.messages.StoredMessage;
+import com.exactpro.cradle.testevents.StoredTestEventWrapper;
+import com.exactpro.cradle.utils.CompressionUtils;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+import java.util.Map;
+import java.util.Optional;
+import java.util.StringJoiner;
+
+public class RecoveryState {
+    private final InnerEvent lastProcessedEvent;
+
+    private final Map<String, InnerMessage> lastProcessedMessages;
+
+    private final long lastNumberOfEvents;
+
+    private final long lastNumberOfMessages;
+
+    private static final ObjectMapper MAPPER = new ObjectMapper().findAndRegisterModules()
+            .configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
+
+    private static final Logger logger = LoggerFactory.getLogger(RecoveryState.class);
+
+    public RecoveryState(@JsonProperty("lastProcessedEvent") InnerEvent lastProcessedEvent,
+                         @JsonProperty("lastProcessedMessages") Map<String, InnerMessage> lastProcessedMessages,
+                         @JsonProperty("lastNumberOfEvents") long lastNumberOfEvents,
+                         @JsonProperty("lastNumberOfMessages") long lastNumberOfMessages) {
+        this.lastProcessedEvent = Optional.ofNullable(lastProcessedEvent).orElse(null);
+        this.lastProcessedMessages = Optional.ofNullable(lastProcessedMessages).orElse(null);
+        this.lastNumberOfEvents = lastNumberOfEvents;
+        this.lastNumberOfMessages = lastNumberOfMessages;
+    }
+
+    public InnerEvent getLastProcessedEvent() { return lastProcessedEvent; }
+
+    public Map<String, InnerMessage> getLastProcessedMessages() { return lastProcessedMessages; }
+
+    public long getLastNumberOfEvents() { return lastNumberOfEvents; }
+
+    public long getLastNumberOfMessages() { return lastNumberOfMessages; }
+
+    public String convertToJson() {
+        try {
+            return MAPPER.writeValueAsString(this);
+        } catch (JsonProcessingException e) {
+            logger.error("Failed to convert recovery state to JSON", e);
+        }
+        return null;
+    }
+
+    public static RecoveryState getStateFromJson(String json) {
+        try {
+            return MAPPER.readValue(json, RecoveryState.class);
+        } catch (JsonProcessingException e) {
+            logger.error("Failed to get recovery state from JSON", e);
```

Comment:
Shouldn't we throw an exception in this case? Looks like it is a critical problem